### PR TITLE
BREAKING: Add LazyList - default importing is now Lazy

### PR DESCRIPTION
**Pull request by patricksnape**
This adds the concept of an Immutable LazyList class. This
is essentially a very simple class whereby the elements are
callables that are assumed to do something *on index*. This is
very useful for importing because we just immediately return
a set of partial import methods that wrap the file paths that
need to be imported. The disadvantage is obviously that if the
file system changes after the list is constructed then the
import will fail. However, the low memory and instant access
(think random access) is a great benefit. To allow for backwards
compatibility there is a new as_generator argument that restores
the previous behaviour.

Finally, there is one major change - now, importers that return
lists of assets will not flatten the overall list - for example,
if you import a GIF, previously the frames would be yielded as
seperate images. Instead, the LazyList will not contain a list
that contains all the frames of the GIF. If this is a serious game
changer then we can add a helper generator that will flatten
the lists. Something that neatly wraps
``list(itertools.chain.from_iterable(assets))``.

@jabooth This one is for you mate!

Breaking:

  - Importers now return `LazyList`
  - Landmark resolvers now receive paths rather than objects

### CODE REVIEW

**Comment by patricksnape** [comment]
Travis is failing because the current scipy release is broken :fearful:

File: menpo/base.py [hunk 1]
```diff
@@ -449,3 +450,79 @@ def _use_parent_doc(self, func, source):
...
+        n_elements : `int`
+            The number of elements in the underlying sequence.
+
+        Returns
+        -------
+        lazy : `LazyList`
+            A LazyList where each element returns the underlying indexable
+            object wrapped by ``f``.
+        """
+        return LazyList([partial(f, i) for i in range(n_elements)])
```
Comment (jabooth) [inline menpo/base.py]: is the motivation for this video sequences? Can see the use case if some API vends lists of callables that take an index arg that could be dropped straight in here, just wondering.

File: menpo/base.py [hunk 2]
```diff
@@ -449,3 +450,79 @@ def _use_parent_doc(self, func, source):
...
+        ----------
+        f : `callable`
+            Callable to wrap each element with.
+
+        Returns
+        -------
+        lazy : `LazyList`
+            A new LazyList where each element is wrapped by ``f``.
+        """
+        return LazyList([partial(f, x) for x in self])
```
Comment (jabooth) [inline menpo/base.py]: <3

File: menpo/io/input/base.py [hunk 3]
```diff
@@ -44,13 +46,13 @@ def data_path_to(asset_filename):
...
 
-def same_name(asset):
+def same_name(path):
     r"""
-    Menpo's default landmark resolver. Returns all landmarks found to have
+    Menpo's default image landmark resolver. Returns all landmarks found to have
     the same stem as the asset.
     """
     # pattern finding all landmarks with the same stem
-    pattern = asset.path.with_suffix('.*')
```
Comment (jabooth) [inline menpo/io/input/base.py]: Don't mind this change, but it is another breaking one, can you note it in the PR notes so we catch it in the changelog?

File: menpo/io/input/base.py [hunk 4]
```diff
@@ -386,21 +410,45 @@ def _import_glob_generator(pattern, extension_map, max_assets=None,
...
+                                  importer_kwargs=importer_kwargs)
+                          for f in filepaths])
 
-    if verbose:
+    if verbose and as_generator:
         # wrap the generator with the progress reporter
-        generator = print_progress(generator, prefix='Importing assets',
+        lazy_list = print_progress(lazy_list, prefix='Importing assets',
                                    n_items=n_files)
+    elif verbose:
```
Comment (jabooth) [inline menpo/io/input/base.py]: nice. :)

**Comment by jabooth** [comment]
@patricksnape reviewed, just haven't tried it yet. Only few minor points but looks great. I'll try it later tonight. Nice work!

File: menpo/base.py (hunk 1 above)
Comment (patricksnape) [inline menpo/base.py]: Yes the video API will use this method

File: menpo/io/input/base.py (hunk 3 above)
Comment (patricksnape) [inline menpo/io/input/base.py]: done

File: menpo/io/input/base.py (hunk 4 above)
Comment (patricksnape) [inline menpo/io/input/base.py]: Happy with that message?

**Comment by patricksnape** [comment]
@jabooth Any thoughts on adding a cached lazy list or not?

**Comment by jabooth** [comment]
Hmm an 'as_cached()' method on LazyList would be sweet:
'''
menpo.io.import_images(...).as_cached()
'''
Implement just as memoise-wrapping 'self._callables' right?

**Comment by patricksnape** [comment]
I was thinking like a kwarg on construction? So if you use choose to cache then it will memoise the accesses like you said. Then the overhead is free.

**Comment by jabooth** [comment]
After discussions with @patricksnape this morning, we are going to hold off on implementing caching for now, see #671 for discussion.

**Comment by jabooth** [comment]
Happy for this to come in when tests pass! +1

**Pull request merged**